### PR TITLE
Add MultiWOZ 2.4 DST evaluation

### PR DESCRIPTION
**Pull request by WeixuanZ**

### CODE REVIEW

File: mwzeval/utils.py [hunk 1]
```diff
@@ -192,3 +201,95 @@ def parse_state(turn):
...
+                        }
+                        dialog_states_of_one_domain.update(new_states)
+                if isinstance(v, str) and v:
+                    slot_name = f"book{k}" if prepend_book else k
+                    dialog_states_of_one_domain[slot_name] = v
+
+            new_states = values["semi"]
+            dialog_states_of_one_domain.update(new_states)
+
+            dialog_states_of_one_domain = {
```
Comment (alexcoca) [inline mwzeval/utils.py]: domain_dial_state ?

Comment (alexcoca) [inline mwzeval/utils.py]: domain_dial_state ?

File: mwzeval/utils.py [hunk 2]
```diff
@@ -192,3 +201,95 @@ def parse_state(turn):
...
+                slot_name: value.split('|')[0]  # use the first value
+                for slot_name, value in dialog_states_of_one_domain.items()
+                if is_filled(value)
```
Comment (alexcoca) [inline mwzeval/utils.py:270]: okay...from what I remember, there may be a catch... can you assert value is not `?`

Comment (alexcoca) [inline mwzeval/utils.py:270]: does the official evaluator on the multiwoz 22 also exclude these `none` and `not mentioned` values?

File: mwzeval/utils.py [hunk 3]
```diff
@@ -192,3 +201,95 @@ def parse_state(turn):
...
+            dialog_states_of_one_domain = {}
+
+            for k, v in values["book"].items():
+                # Note: "booked" is not really a state, just booking confirmation
+                if k == 'booked':
+                    continue
+                if isinstance(v, list):
+                    for item_dict in v:
+                        new_states = {
+                            (f"book{slot_name}" if prepend_book else slot_name): slot_val
```
Comment (alexcoca) [inline mwzeval/utils.py:256]: is this`bookpeople` or something like that in the corpus? or are we missing some dash etc?

File: mwzeval/utils.py [hunk 4]
```diff
@@ -192,3 +201,95 @@ def parse_state(turn):
...
+                        dialog_states_of_one_domain.update(new_states)
+                if isinstance(v, str) and v:
+                    slot_name = f"book{k}" if prepend_book else k
+                    dialog_states_of_one_domain[slot_name] = v
+
+            new_states = values["semi"]
+            dialog_states_of_one_domain.update(new_states)
+
+            dialog_states_of_one_domain = {
+                slot_name: value.split('|')[0]  # use the first value
```
Comment (alexcoca) [inline mwzeval/utils.py]: let's make sure there aren't any other symbols that indicate multiple vals in the strings, check Google's proc code!

**Review by alexcoca — COMMENTED** [review]
Okay, looks roughly fine to me. I'll give it another look later

File: mwzeval/utils.py (hunk 2 above)
Comment (WeixuanZ) [reply mwzeval/utils.py:270]: I don't think these slot values exist in MultiWOZ 2.2 dataset

File: mwzeval/utils.py (hunk 3 above)
Comment (WeixuanZ) [reply mwzeval/utils.py:256]: MultiWOZ 2.2 appends `book` to booking slots, e.g. in dialogue `PMUL2596`

MultiWOZ 2.2:
```
          {
            "actions": [],
            "service": "hotel",
            "slots": [],
            "state": {
              "active_intent": "book_hotel",
              "requested_slots": [],
              "slot_values": {
                "hotel-bookpeople": [
                  "4"
                ],
                "hotel-bookstay": [
                  "4"
                ],
                "hotel-name": [
                  "arbury lodge guesthouse"
                ]
              }
            }
          },
```

MultiWOZ 2.4:
```
                    "hotel": {
                        "book": {
                            "booked": [],
                            "stay": "4",
                            "day": "",
                            "people": "4"
                        },
                        "semi": {
                            "name": "arbury lodge guesthouse",
                            "area": "not mentioned",
                            "parking": "not mentioned",
                            "pricerange": "not mentioned",
                            "stars": "not mentioned",
                            "internet": "not mentioned",
                            "type": "not mentioned"
                        }
                    },
```

I don't think these prefixes are necessary for 2.4, and Google D3ST preprocessing script (https://github.com/google-research/task-oriented-dialogue/blob/main/state_tracking/utils/multiwoz_utils.py) actually removes the `book` prefix.

**Comment by WeixuanZ** [comment]
Thanks @alexcoca!

File: mwzeval/utils.py (hunk 4 above)
Comment (WeixuanZ) [reply mwzeval/utils.py]: Good point, indeed there are! Will fix

Comment (WeixuanZ) [reply mwzeval/utils.py]: Interestingly, the Google MultiWOZ 2.1 -> SGD conversion script (https://github.com/google-research/google-research/blob/master/schema_guided_dst/multiwoz/create_data_from_multiwoz.py) doesn't seem to handle `<` and `>`.

File: mwzeval/normalization.py [hunk 5]
```diff
@@ -78,6 +78,8 @@ def type_to_canonical(type_string):
             return "nightclub"
         elif type_string == "guest house":
             return "guesthouse"
+        elif type_string == "concert hall":
```
Comment (alexcoca) [inline mwzeval/normalization.py:81]: are we extending the normalisation code from the official repo here?

File: mwzeval/utils.py (hunk 3 above)
Comment (alexcoca) [reply mwzeval/utils.py:256]: crazy how complicated this became, it's ridiculous.

**Review by alexcoca — COMMENTED** [review]
Good work, let's discuss changes to normalisation.

File: mwzeval/utils.py [hunk 6]
```diff
@@ -192,3 +201,95 @@ def parse_state(turn):
         json.dump(states, f, indent=2)
 
     return references, states
+
+
+def load_multiwoz24():
+    def is_filled(slot_value: str) -> bool:
+        """Whether a slot value is filled."""
+        slot_value = slot_value.lower()
+        return slot_value and slot_value != "not mentioned" and slot_value != "none"
```
Comment (alexcoca) [inline mwzeval/utils.py:214]: I'm not too sure about `none` semantics in the corpus ... when PRing upstream, ask the maintainers their opinion on this? I presume that 22 only has active slots in the frames and the point of this function is to achieve the same?

Comment (alexcoca) [reply mwzeval/utils.py:214]: tag MultiWOZ 2.4 paper authors as well asking for their opinion (if you can find them on github)

File: mwzeval/utils.py (hunk 4 above)
Comment (alexcoca) [reply mwzeval/utils.py]: oh well :) you always find new things in MultiWOZ, can spend a lifetime on this alone.

**Pull request merged**